### PR TITLE
Track caml_alloc_custom_mem allocations with Statmemprof

### PR DESCRIPTION
**Pull request by stedolan**
Memory allocated via caml_alloc_custom_mem (e.g. Bigarrays, I/O channels) is managed by the GC, despite being stored outside the GC heap. However, these allocations are currently invisible to Statmemprof, which makes it tricky to profile some applications that heavily use bigarrays.

This patch makes these allocations visible to Statmemprof.

Note: Allocating a bigarray actually involves two separate allocations. One is a small Custom_tag block on the GC heap, which points to the (usually larger) second allocation, a malloc'd block containing the actual data. With this patch, these are *both* tracked, and it's possible to get two separate Statmemprof events from a call to caml_alloc_custom, if both allocations happen to be sampled. (I think this is the correct behaviour, but it's worth pointing out explicitly).

cc @jhjourdan

### CODE REVIEW

**Comment by lpw25** [comment]
I think I would expect the off-heap allocations to be marked in some way in the API. Maybe their own `alloc_custom` field in `tracker`. Maybe a `bool` field in `allocation`.

**Comment by stedolan** [comment]
I think the most natural way to do that would be to change `unmarshalled : bool` in `allocation` to ```source : [`Normal | `Marshal | `Custom]``` or similar.

I'm a bit ambivalent about this, though. It's a breaking API change and the information isn't particularly useful: the backtrace generally tells you what you want to know.

**Comment by lpw25** [comment]
> It's a breaking API change

Note that the `Memprof` module currently has: `Note: this API is EXPERIMENTAL. It may change without prior notice.` in its documentation.

File: runtime/memprof.c [hunk 1]
```diff
@@ -765,6 +765,25 @@ void caml_memprof_track_alloc_shr(value block)
...
+  n_samples = rand_binom(Wsize_bsize(bytes));
+  if (n_samples == 0) return;
+
+  callstack = capture_callstack_postponed();
+  if (callstack == 0) return;
+
+  new_tracked(n_samples, Wsize_bsize(bytes), 0, Is_young(block),
+              block, callstack);
+  check_action_pending();
+}
```
Comment (gasche) [inline runtime/memprof.c:782]: There is now a fair amount of duplication between `caml_memprof_track_shr` and `caml_memprof_track_custom`, would it make sense to share code?

```c
void caml_memprof_track(size, is_unmarshalled, is_young, block)
{
  uintnat n_samples;
  value callstack = 0;

  if (lambda == 0 || local->suspended) return;

  n_samples = rand_binom(size);
  if (n_samples == 0) return;

  callstack = capture_callstack_postponed();
  if (callstack == 0) return;

  new_tracked(n_samples, size, is_unmarshalled, is_young, block, callstack);
  check_action_pending();
}
```

**Comment by stedolan** [comment]
> Note that the Memprof module currently has: Note: this API is EXPERIMENTAL. It may change without prior notice. in its documentation.

I hadn't realised that was still there! My qualms about API changes have evaporated, I'll add something about custom blocks.

**Comment by jhjourdan** [comment]
Thanks for the PR ! Generally it looks good. I do agree with @lpw25 and @gasche remarks.

I have a few other concerns, though (no strong opinion on either one):

- We might consider doing something with blocks allocated with the old API `caml_alloc_custom`, both because perhaps not all C bindings have done the transition (`caml_alloc_custom_mem` is a relatively new API), and because non-memory resources allocated using `caml_alloc_custom` may also be worth tracking.

- I am not so sure that making custom tracked blocks go though the usual minor -> promotion -> deallocation pipeline is the way to go. After all, we are not tracking heap resources here. I would rather be in favor of using dedicated callbacks for allocating and deallocating out-of-heap resources.

File: runtime/memprof.c (hunk 1 above)
Comment (stedolan) [reply runtime/memprof.c:782]: Done. Turns out there was a third occurrence of more-or-less the same code as well.

**Comment by stedolan** [comment]
> We might consider doing something with blocks allocated with the old API caml_alloc_custom, both because perhaps not all C bindings have done the transition (caml_alloc_custom_mem is a relatively new API), and because non-memory resources allocated using caml_alloc_custom may also be worth tracking.

I did think about this, but I don't think there's anything very sensible we can do. The sampling rate is in units of samples / word, and caml_alloc_custom allocates in some other unknown unit. At what rate should they be sampled? We can't magic up a conversion rate between, say, bytes and filehandles.

> I am not so sure that making custom tracked blocks go though the usual minor -> promotion -> deallocation pipeline is the way to go. After all, we are not tracking heap resources here. I would rather be in favor of using dedicated callbacks for allocating and deallocating out-of-heap resources.

If we were tracking general uses of caml_alloc_custom then I'd agree, but for caml_alloc_custom_mem I don't. Memory allocated by caml_alloc_custom_mem is genuinely GC-managed memory, despite being allocated with a different allocator. Its lifetime is determined by the garbage collector following the same minor/promote/major lifecycle as normal allocations.

**Comment by gasche** [comment]
The prototype of `caml_alloc_custom` is as follows:

```c
CAMLextern value caml_alloc_custom(struct custom_operations * ops,
                                   uintnat size, /*size in bytes*/
                                   mlsize_t mem, /*resources consumed*/
                                   mlsize_t max  /*max resources*/);
```

if I understand correctly (see also the [manual](https://caml.inria.fr/pub/docs/manual-ocaml-4.11/intfc.html#ss:c-custom-alloc)), `mem` and `max` describe an abstract "resource cost" `mem/max` for the allocation.

If the current sampling rate is `lambda` (with 0 <= lambda <= 1), I would expect the average number of samples for such an allocation to be `lambda * mem / max`.

**Comment by lpw25** [comment]
Could I suggest leaving `caml_alloc_custom` out of this PR? I think that is a much more contentious change. My experience is that the `mem` and `max` parameters that users choose are usually nonsense, and I'm not keen on us relying on them for anything. Besides, I think most code now just uses the values that mean "there are infinite resources".

**Comment by gasche** [comment]
(Yes, personally I'm happy to leave the general/abstract resources out of scope of this PR.)

**Comment by stedolan** [comment]
I also agree it's best to leave this out for now, but I'll leave some notes here in case we come back to it:

> if I understand correctly (see also the manual), mem and max describe an abstract "resource cost" mem/max for the allocation.
> 
> If the current sampling rate is lambda (with 0 <= lambda <= 1), I would expect the average number of samples for such an allocation to be lambda * mem / max.

The resource costs mem/max are used to speed up the GC to ensure that a major GC cycles complete before resources are exhausted. The idea is that by the time the mem/max values sum to 1, a full GC cycle will have been completed. So, using lambda * mem / max as the sampling rate would end up with an average of lambda samples *per GC cycle* rather than lambda samples *per word allocated* (The number of words allocated during a GC cycle varies according to the heap size, but could be billions with a large heap). If we were to use this, it'd have to be something like lambda * words_allocated_per_cycle * mem / max to make the units work out.

Having said that, I agree with Leo that in practice the mem/max ratio is usually nonsense, so I'm also not keen on using these numbers.

**Comment by gasche** [comment]
Ah, yes, sorry. Yes, `words_allocated_per_cycle` sounds like a quantity we can measure and approximate reasonably enough to make the computation work out.

I don't think there is a good solution if we assume that the information we are given is completely incorrect. Maybe we could have a runtime setting to ask to ignore it, and in this case make a default choice; but by default we should probably use it, and encourage people to fix their programs/libraries -- or redesign the API!

**Comment by jhjourdan** [comment]
Alright, I reviewed the current version of the code. It looks good to me, you can merge this.

It is important that the client code knows that some allocated blocks are custom, since:
- This memory counts for the size of the process, but not for the size of the heap
- A word allocated in the heap costs more than a word in terms of process memory, since the GC needs a fraction of free space in the heap. This is not the case for out-of-heap memory.


> > We might consider doing something with blocks allocated with the old API caml_alloc_custom, both because perhaps not all C bindings have done the transition (caml_alloc_custom_mem is a relatively new API), and because non-memory resources allocated using caml_alloc_custom may also be worth tracking.
> 
> I did think about this, but I don't think there's anything very sensible we can do. The sampling rate is in units of samples / word, and caml_alloc_custom allocates in some other unknown unit. At what rate should they be sampled? We can't magic up a conversion rate between, say, bytes and filehandles.

Alright. There might be something to do, but let's wait when we have a clear use case which will guide us.

> > I am not so sure that making custom tracked blocks go though the usual minor -> promotion -> deallocation pipeline is the way to go. After all, we are not tracking heap resources here. I would rather be in favor of using dedicated callbacks for allocating and deallocating out-of-heap resources.
> 
> If we were tracking general uses of caml_alloc_custom then I'd agree, but for caml_alloc_custom_mem I don't. Memory allocated by caml_alloc_custom_mem is genuinely GC-managed memory, despite being allocated with a different allocator. Its lifetime is determined by the garbage collector following the same minor/promote/major lifecycle as normal allocations.

Ok, you convinced me.

**Comment by stedolan** [comment]
Any objections to merging this?

**Comment by jhjourdan** [comment]
No. Merging.

**Comment by hannesm** [comment]
Any chance this being part of the 4.12 release?
I cherry-picked it locally, and it cleanly applies to the 4.12 branch. It fixes the issue of "bigarray allocations are not tracked by statmemprof". it would be (from my point of view) a low-risk addition to the 4.12 branch.

Thanks for your consideration (and for the patch in the first place).

**Comment by stedolan** [comment]
> Any chance this being part of the 4.12 release?
I cherry-picked it locally, and it cleanly applies to the 4.12 branch. It fixes the issue of "bigarray allocations are not tracked by statmemprof". it would be (from my point of view) a low-risk addition to the 4.12 branch.

Makes sense to me. I think it's up to @Octachron to decide whether it's small / bugfix-y enough to include in the release.

**Comment by Octachron** [comment]
Fixing a blind spot in an experimental API sounds like a relatively safe bugfix indeed.

**Comment by hannesm** [comment]
@stedolan @Octachron thanks for your positive replies, would be great if someone could do the actual cherry-pick to the 4.12 branch.

**Comment by stedolan** [comment]
Done, cherry-picked in 86fe55f8b

**Comment by kit-ty-kate** [comment]
Ported to batteries https://github.com/ocaml-batteries-team/batteries-included/pull/994
I don't see any other packages that fail because of that at the moment but I might have missed one

**Comment by kit-ty-kate** [comment]
just for next time, if there are breaking changes in the API between two alphas/betas I'd be very happy to be notified to propagate the changes downstream.

**Pull request merged**